File: backend/app/schemas/tax.py
from __future__ import annotations

from datetime import datetime, timezone
from decimal import Decimal
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field, field_validator


class FilingStatus(StrEnum):
    SINGLE = "single"
    MARRIED_JOINTLY = "married_filing_jointly"
    MARRIED_SEPARATELY = "married_filing_separately"
    HEAD_OF_HOUSEHOLD = "head_of_household"


class WorkflowStatus(StrEnum):
    UPLOADED = "uploaded"
    PARSING = "parsing"
    CALCULATING = "calculating"
    VERIFYING = "verifying"
    REMEDIATING = "remediating"
    COMPLETED = "completed"
    MANUAL_REVIEW = "manual_review"
    FAILED = "failed"


class SourceEvidence(BaseModel):
    field: str
    page: int = 1
    raw_text: str = ""
    source: str = "ocr"
    confidence: float = Field(default=0.0, ge=0, le=1)


class TaxpayerData(BaseModel):
    employee_name: str = ""
    employer_name: str = ""
    filing_status: FilingStatus = FilingStatus.SINGLE
    tax_year: int = 2025
    wages: Decimal = Decimal("0")
    federal_tax_withheld: Decimal = Decimal("0")
    state_tax_withheld: Decimal = Decimal("0")
    other_income: Decimal = Decimal("0")
    itemized_deductions: Decimal = Decimal("0")
    state_tax_rate: Decimal = Decimal("0.05")
    state: str = ""
    evidence: list[SourceEvidence] = Field(default_factory=list)
    field_confidence: dict[str, float] = Field(default_factory=dict)

    @field_validator(
        "wages",
        "federal_tax_withheld",
        "state_tax_withheld",
        "other_income",
        "itemized_deductions",
        "state_tax_rate",
        mode="before",
    )
    @classmethod
    def normalize_decimal(cls, value: Any) -> Decimal:
        if value in (None, ""):
            return Decimal("0")
        if isinstance(value, str):
            value = value.replace("$", "").replace(",", "").strip()
        return Decimal(str(value))


class TaxCalculation(BaseModel):
    gross_income: Decimal
    standard_deduction: Decimal
    deductions: Decimal
    taxable_income: Decimal
    federal_tax: Decimal
    state_tax: Decimal
    total_tax: Decimal
    total_withholding: Decimal
    refund: Decimal
    tax_due: Decimal
    trace: list[str] = Field(default_factory=list)


class VerificationCheck(BaseModel):
    name: str
    passed: bool
    message: str
    weight: float = 1.0


class VerificationResult(BaseModel):
    valid: bool
    confidence_score: float = Field(ge=0, le=1)
    checks: list[VerificationCheck]
    hallucination_flags: list[str] = Field(default_factory=list)
    errors: list[str] = Field(default_factory=list)


class AuditEntry(BaseModel):
    agent: str
    action: str
    reason: str
    details: dict[str, Any] = Field(default_factory=dict)
    timestamp: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )


class FilingReceipt(BaseModel):
    submission_id: str
    reference_number: str
    timestamp: datetime
    filing_status: str


class SubmissionResult(BaseModel):
    submission_id: str
    status: WorkflowStatus
    original_filename: str
    extracted_data: TaxpayerData | None = None
    calculation: TaxCalculation | None = None
    verification: VerificationResult | None = None
    audit_trail: list[AuditEntry] = Field(default_factory=list)
    receipt: FilingReceipt | None = None
    report_url: str | None = None
    error: str | None = None

File: frontend/src/types/tax.ts
export type WorkflowStatus =
  | "uploaded"
  | "parsing"
  | "calculating"
  | "verifying"
  | "remediating"
  | "completed"
  | "manual_review"
  | "failed";

export interface AuditEntry {
  agent: string;
  action: string;
  reason: string;
  timestamp: string;
  details: Record<string, unknown>;
}

export interface SubmissionResult {
  submission_id: string;
  status: WorkflowStatus;
  original_filename: string;
  extracted_data?: {
    employee_name: string;
    employer_name: string;
    filing_status: string;
    tax_year: number;
    wages: string;
    federal_tax_withheld: string;
    state_tax_withheld: string;
  };
  calculation?: {
    gross_income: string;
    total_income?: string;
    adjustments?: string;
    adjusted_gross_income?: string;
    deductions: string;
    qbi_deduction?: string;
    taxable_income: string;
    income_tax_before_credits?: string;
    nonrefundable_credits?: string;
    other_taxes?: string;
    federal_tax: string;
    state_tax: string;
    child_tax_credit?: string;
    earned_income_credit?: string;
    self_employment_tax?: string;
    total_payments?: string;
    tax_table_used?: boolean;
    params_verified?: boolean;
    refund: string;
    tax_due: string;
    state_refund?: string;
    state_balance_due?: string;
  };
  verification?: {
    valid: boolean;
    confidence_score: number;
    correctness_ok?: boolean;
    completeness_ok?: boolean;
    errors: string[];
    hallucination_flags: string[];
    checks?: { name: string; passed: boolean; message: string }[];
  };
  audit_trail: AuditEntry[];
  receipt?: {
    reference_number: string;
    filing_status: string;
  };
  report_url?: string;
  error?: string;
}
